pkg/storegateway/indexheader/index/symbols.go

```diff
@@ -100,6 +100,67 @@ func (s *Symbols) Lookup(o uint32) (sym string, err error) {
 	return sym, nil
 }
 
+func (s *Symbols) Reader() *SymbolsTableReader {
+	d := s.factory.NewDecbufAtUnchecked(s.tableOffset)
+	d.ResetAt(s.offsets[0])
+	return &SymbolsTableReader{
+		d: &d,
+		s: s,
+	}
+}
+
+type SymbolsTableReader struct {
+	// atSymbol is the sequence of the symbol in V2, and the offset of the symbol in V1
+	atSymbol uint32
+	d        *streamencoding.Decbuf
+	s        *Symbols
+}
+
+func (r *SymbolsTableReader) Close() error {
+	return r.d.Close()
+}
+
+func (r *SymbolsTableReader) Read(o uint32) (string, error) {
```

Comment:
Given the implementation differs significantly for v1 and v2, should we define an interface for `SymbolsTableReader` and have different implementations for v1 and v2, similar to what we do for `PostingsOffsetTable`?

Reply:
done in https://github.com/grafana/mimir/pull/5073/commits/f74ae4451bdd73a22c9406879d262fd6aa44c65b